### DISCUSSION TITLE
New Sign/verify APIS

### DISCUSSION THREAD

**Discussion by slontis** [openssl/openssl#22671] [General Discussion]
The composite algorithm design 
(See https://github.com/openssl/openssl/blob/master/doc/designs/fetching-composite-algorithms.md)
lists the following interfaces for signatures.

````
EVP_DigestSignInit_ex2(EVP_PKEY_CTX **pctx,
                       EVP_SIGNATURE *sig, EVP_PKEY *pkey,
                       OSSL_LIB_CTX *libctx, const OSSL_PARAM params[]);
EVP_DigestVerifyInit_ex2(EVP_PKEY_CTX **pctx,
                         EVP_SIGNATURE *sig, EVP_PKEY *pkey,
                         OSSL_LIB_CTX *libctx, const OSSL_PARAM params[]);
````

Note that this would then also need a different update() and final() and/or one shot() in order to pass the PKEY_CTX not the EVP_MD_CTX?

There has been a mention of using the EVP_PKEY_sign() interface for HSS but I think maybe it should look more like the above interface (in terms of parameters, not necessarily the name).

@levitte Would there be much of a problem just passing a EVP_PKEY_CTX * in the above DigestSignInit() rather than resolving it internally. i.e. the user creates the pkey_ctx. (It would of course have to figure out mismatches between the key and the signature but that would be doable).

For HSS whatever interface we come up with above it is not much of a stretch to go from that to an init function that also passes the signature and signature length in the init. It could even be just the one function in that case that allows the params to be NULL/0. The final could remain the same and also be able to handle NULL/0. A flag would then be needed internally to manage when the signature was set (init or final). I am happy that the digest is not in the API since HSS and XMSS have that information embedded in the types that they use (that exist inside both the key data and the signature data).

**Comment by levitte** [comment]
I went along with the idea to extend `EVP_PKEY_sign()` and `EVP_PKEY_verify()` for this sort of thing, see #22672

**Comment by levitte** [comment]
With an explicitly fetched `EVP_SIGNATURE`, the provider can offer whatever it needs under whatever algorithm name (and OID) that is appropriate, as long as it fits the expectations of the rest of the functionality.  It sounds like HSS and XMSS would be possible to support under those conditions (and given the expanded functionality proposed in #22672)...  or am I missing something?

**Reply by slontis** [reply]
No your approach looks good to me.
